ADBDEV-4345: Fix wrong allocation of writer slice in case of modifying shared CTE

trexxet2 opened the pull request:
Fix wrong allocation of writer slice in case of modifying shared CTE

Initially, when the CTE was first encountered, the create_ctescan_plan()
function constructed a subplan, and its slice was assigned the
GANGTYPE_PRIMARY_WRITER gang type. However, for subsequent occurrences of the
CTE, the subplan construction was bypassed. If these occurrences were located in
different slices and the CTE involved DML, their slices would sometimes be
assigned the GANGTYPE_PRIMARY_WRITER gang type. This lead to issues when the
ShareInputScans were assigned producer and consumer roles by the
apply_shareinput_dag_to_tree(), which uses a specific tree traversal order, and
in some cases, there was a scenario when the consumer was assigned to a
GANGTYPE_PRIMARY_WRITER slice, while the producer was placed in a
GANGTYPE_PRIMARY_READER slice, resulting in an execution error when the reader
gang attempted to execute a ModifyTable node.

To resolve this issue, the patch introduces a check for the gang type of
root->curSlice when creating the shared plan for the first time. If the gang
type has changed compared to its previous value during planning, the
ShareInputScan is marked as a writer slice maker by setting the
rootSliceIsWriter flag to true. This state is also applied to all subsequent
occurrences of the CTE. After assigning consumer and producer roles to each
ShareInputScan, the correct gang type is set for the producer
(GANGTYPE_PRIMARY_WRITER) and for the consumers (GANGTYPE_PRIMARY_READER, if the
consumer is in a different slice from the producer) within the
shareinput_mutator_xslice_2() function. This ensures that the gang type is
correctly assigned, preventing the execution error encountered in the original
scenario.

Co-authored-by: Alexander Kondakov \<a.kondakov@arenadata.io\>

deart2k committed:
ADBDEV-3405 Provide infrastructure to build gpdb 7 with make dist

deart2k committed:
Add altlinux and astra linux to makefile

Commit:
Fix wrong writer slice on mod shared CTE

bimboterminator1 commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -1014,6 +1014,12 @@ shareinput_mutator_xslice_2(Node *node, PlannerInfo *root, bool fPop)
+		/*
+		 * Remove writer gang for the shareinput scan readers
+		 */
+		if (currentSlice->gangType == GANGTYPE_PRIMARY_WRITER && !plan->lefttree)
+			currentSlice->gangType = GANGTYPE_PRIMARY_READER;
```
@trexxet2, proposed solution raises doubts.
First of all, about this fragment of code. Why do we think that SISC without subplan, when being under writing slice, **always corresponds to the modifying CTE**? I guess, it's possible that GANGTYPE_PRIMARY_WRITER related to other operation, not to the current SISC. And in this case we just overwrite correct `gangType`. 
I manged to find such case:
**Wihtout patch**:
```
set gp_cte_sharing=1;
set optimizer = 0;
create table t_hashed (i int, j int);
create table t_hashed2 (i int, j int);
 explain (slicetable, costs off) with cte1 as (select * from t_hashed)  
update t_hashed2 set j=a.i from cte1 a join (select x.i, x.j from cte1 x limit 1) b on a.i=b.i where b.i=t_hashed2.i;
                                          QUERY PLAN                                          
----------------------------------------------------------------------------------------------
 Update on t_hashed2
   ->  Hash Join
         Hash Cond: (a.i = b.i)
         ->  Subquery Scan on a
               ->  Shared Scan (share slice:id 0:0)
         ->  Hash
               ->  Hash Join
                     Hash Cond: (t_hashed2.i = b.i)
                     ->  Seq Scan on t_hashed2
                     ->  Hash
                           ->  Redistribute Motion 1:3  (slice1; segments: 1)
                                 Hash Key: b.i
                                 ->  Subquery Scan on b
                                       ->  Limit
                                             ->  Gather Motion 3:1  (slice2; segments: 3)
                                                   ->  Limit
                                                         ->  Shared Scan (share slice:id 2:0)
                                                               ->  Seq Scan on t_hashed
 Optimizer: Postgres-based planner
 Slice 0: Primary Writer; root 0; parent -1; gang size 3
 Slice 1: Singleton Reader; root 0; parent 0; gang size 1; segment 1
 Slice 2: Reader; root 0; parent 1; gang size 3
(22 rows)
```
First Shared scan is under the writing slice, however it's a consumer.
**With patch**
```
explain (slicetable, costs off) with cte1 as (select * from t_hashed) 
 update t_hashed2 set j=a.i from cte1 a join (select x.i, x.j from cte1 x limit 1) b on a.i=b.i where b.i=t_hashed2.i;
                                          QUERY PLAN                                          
----------------------------------------------------------------------------------------------
 Update on t_hashed2
   ->  Hash Join
         Hash Cond: (a.i = b.i)
         ->  Subquery Scan on a
               ->  Shared Scan (share slice:id 0:0)
         ->  Hash
               ->  Hash Join
                     Hash Cond: (t_hashed2.i = b.i)
                     ->  Seq Scan on t_hashed2
                     ->  Hash
                           ->  Redistribute Motion 1:3  (slice1; segments: 1)
                                 Hash Key: b.i
                                 ->  Subquery Scan on b
                                       ->  Limit
                                             ->  Gather Motion 3:1  (slice2; segments: 3)
                                                   ->  Limit
                                                         ->  Shared Scan (share slice:id 2:0)
                                                               ->  Seq Scan on t_hashed
 Optimizer: Postgres-based planner
 Slice 0: Reader; root 0; parent -1; gang size 3
 Slice 1: Singleton Reader; root 0; parent 0; gang size 1; segment 1
 Slice 2: Primary Writer; root 0; parent 1; gang size 3
(22 rows)
```

Commit:
check hasModifyingCTE as well

bimboterminator1 commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -1014,6 +1014,17 @@ shareinput_mutator_xslice_2(Node *node, PlannerInfo *root, bool fPop)
+		 * Remove writer gang for the shareinput scan readers in case
+		 * of modifying CTE
+		 */
+		if (root->parse->hasModifyingCTE &&
+			currentSlice->gangType == GANGTYPE_PRIMARY_WRITER
+			&& !plan->lefttree)
+		{
+			currentSlice->gangType = GANGTYPE_PRIMARY_READER;
+		}
```
One more error
**Without patch**
```
create table t_hashed(i int, j int);
create table t2 (i int, j int);


explain (slicetable,costs off) with cte as (insert into t_hashed select i, i*2 from generate_series(1,5) i  returning *) 
select * from t2 join (select * from cte union all select * from cte) x  using (i);
                                              QUERY PLAN                                              
------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)
   ->  Hash Join
         Hash Cond: (t2.i = "*SELECT* 1".i)
         ->  Seq Scan on t2
         ->  Hash
               ->  Redistribute Motion 3:3  (slice2; segments: 3)
                     Hash Key: "*SELECT* 1".i
                     ->  Append
                           ->  Subquery Scan on "*SELECT* 1"
                                 ->  Subquery Scan on cte
                                       ->  Shared Scan (share slice:id 2:0)
                                             ->  Insert on t_hashed
                                                   ->  Redistribute Motion 1:3  (slice3; segments: 1)
                                                         Hash Key: i.i
                                                         ->  Function Scan on generate_series i
                           ->  Subquery Scan on "*SELECT* 2"
                                 ->  Subquery Scan on cte_1
                                       ->  Shared Scan (share slice:id 2:0)
 Optimizer: Postgres-based planner
 Slice 0: Dispatcher; root 0; parent -1; gang size 0
 Slice 1: Reader; root 0; parent 0; gang size 3
 Slice 2: Primary Writer; root 0; parent 1; gang size 3
 Slice 3: Singleton Reader; root 0; parent 2; gang size 1; segment 1
(23 rows)
```

**With patch**
```
explain (slicetable,costs off) with cte as (insert into t_hashed select i, i*2 from generate_series(1,5) i  returning *)
 select * from t2 join (select * from cte union all  select * from cte) x  using (i);
                                              QUERY PLAN                                              
------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)
   ->  Hash Join
         Hash Cond: (t2.i = "*SELECT* 1".i)
         ->  Seq Scan on t2
         ->  Hash
               ->  Redistribute Motion 3:3  (slice2; segments: 3)
                     Hash Key: "*SELECT* 1".i
                     ->  Append
                           ->  Subquery Scan on "*SELECT* 1"
                                 ->  Subquery Scan on cte
                                       ->  Shared Scan (share slice:id 2:0)
                                             ->  Insert on t_hashed
                                                   ->  Redistribute Motion 1:3  (slice3; segments: 1)
                                                         Hash Key: i.i
                                                         ->  Function Scan on generate_series i
                           ->  Subquery Scan on "*SELECT* 2"
                                 ->  Subquery Scan on cte_1
                                       ->  Shared Scan (share slice:id 2:0)
 Optimizer: Postgres-based planner
 Slice 0: Dispatcher; root 0; parent -1; gang size 0
 Slice 1: Reader; root 0; parent 0; gang size 3
 Slice 2: Reader; root 0; parent 1; gang size 3
 Slice 3: Singleton Reader; root 0; parent 2; gang size 1; segment 1
(23 rows)
```

trexxet2 replied:
Figured out that I erroneously supposed that `shareinput_mutator_xslice_2()` works only at cross slices

Commit:
check cross_slice

bimboterminator1 commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -1014,6 +1014,18 @@ shareinput_mutator_xslice_2(Node *node, PlannerInfo *root, bool fPop)
+		 * Remove writer gang for the shareinput scan readers in case
+		 * of cross-slice modifying CTE
+		 */
+		if (sisc->cross_slice &&
+			root->parse->hasModifyingCTE &&
+			currentSlice->gangType == GANGTYPE_PRIMARY_WRITER &&
+			!plan->lefttree)
+		{
+			currentSlice->gangType = GANGTYPE_PRIMARY_READER;
+		}
```
Well, there is a case: when in **one slice we have producer and consumer**, and the **third** consumer is in another slice. In this case the first consumer will have `sisc->cross_slice` `true`, and it will overwritte the correct writer gangtype.
Before patch:
```
create table t_hashed(i int, j int);
create table t2 (i int, j int);

explain (slicetable,costs off) with cte as (insert into t_hashed select i, i*2 from generate_series(1,5) i  returning *) 
select * from cte union all select * from cte union all (select t2.j, cte.i from cte join t2 using (i));
                                  QUERY PLAN                                  
------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)
   ->  Append
         ->  Subquery Scan on cte
               ->  Shared Scan (share slice:id 1:0)
                     ->  Insert on t_hashed
                           ->  Redistribute Motion 1:3  (slice2; segments: 1)
                                 Hash Key: i.i
                                 ->  Function Scan on generate_series i
         ->  Subquery Scan on cte_1
               ->  Shared Scan (share slice:id 1:0)
         ->  Hash Join
               Hash Cond: (t2.i = cte_2.i)
               ->  Seq Scan on t2
               ->  Hash
                     ->  Redistribute Motion 3:3  (slice3; segments: 3)
                           Hash Key: cte_2.i
                           ->  Subquery Scan on cte_2
                                 ->  Shared Scan (share slice:id 3:0)
 Optimizer: Postgres-based planner
 Slice 0: Dispatcher; root 0; parent -1; gang size 0
 Slice 1: Primary Writer; root 0; parent 0; gang size 3
 Slice 2: Singleton Reader; root 0; parent 1; gang size 1; segment 1
 Slice 3: Reader; root 0; parent 1; gang size 3
(23 rows)
```
With patch:
```
explain (slicetable,costs off) with cte as (insert into t_hashed select i, i*2 from generate_series(1,5) i  returning *) 
select * from cte union all select * from cte union all (select t2.j, cte.i from cte join t2 using (i));
                                  QUERY PLAN                                  
------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)
   ->  Append
         ->  Subquery Scan on cte
               ->  Shared Scan (share slice:id 1:0)
                     ->  Insert on t_hashed
                           ->  Redistribute Motion 1:3  (slice2; segments: 1)
                                 Hash Key: i.i
                                 ->  Function Scan on generate_series i
         ->  Subquery Scan on cte_1
               ->  Shared Scan (share slice:id 1:0)
         ->  Hash Join
               Hash Cond: (t2.i = cte_2.i)
               ->  Seq Scan on t2
               ->  Hash
                     ->  Redistribute Motion 3:3  (slice3; segments: 3)
                           Hash Key: cte_2.i
                           ->  Subquery Scan on cte_2
                                 ->  Shared Scan (share slice:id 3:0)
 Optimizer: Postgres-based planner
 Slice 0: Dispatcher; root 0; parent -1; gang size 0
 Slice 1: Reader; root 0; parent 0; gang size 3
 Slice 2: Singleton Reader; root 0; parent 1; gang size 1; segment 0
 Slice 3: Reader; root 0; parent 1; gang size 3
(23 rows)
```
**I think the most valid solution is to ensure that current SISC consumer is definetly in a different slice from  the producer.**

bimboterminator1 replied:
Another moment is that `hadModifyingCTE` can relate to different CTE, and current SISC consumer can also ruin the gangType
**Without patch:**
```
create table t_repl (i int, j int) distributed replicated;
explain (slicetable,costs off) with cte as (insert into t_hashed select i, i*2 from generate_series(1,5) i  returning *),
cte2 as (select * from t_repl)
select * from cte join cte2 a on cte.i=a.i join (select t2.i from cte2 b join t2 using (i) limit 1) x on x.i=cte.i;
                                          QUERY PLAN                                           
-----------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)
   ->  Hash Join
         Hash Cond: (share0_ref2.i = t_hashed.i)
         ->  Shared Scan (share slice:id 1:0)
         ->  Hash
               ->  Hash Join
                     Hash Cond: (t_hashed.i = t2.i)
                     ->  Insert on t_hashed
                           ->  Redistribute Motion 1:3  (slice2; segments: 1)
                                 Hash Key: i.i
                                 ->  Function Scan on generate_series i
                     ->  Hash
                           ->  Broadcast Motion 1:3  (slice3; segments: 1)
                                 ->  Limit
                                       ->  Gather Motion 3:1  (slice4; segments: 3)
                                             ->  Limit
                                                   ->  Hash Join
                                                         Hash Cond: (share0_ref1.i = t2.i)
                                                         ->  Shared Scan (share slice:id 4:0)
                                                               ->  Seq Scan on t_repl t_repl_1
                                                         ->  Hash
                                                               ->  Seq Scan on t2
 Optimizer: Postgres-based planner
 Slice 0: Dispatcher; root 0; parent -1; gang size 0
 Slice 1: Primary Writer; root 0; parent 0; gang size 3
 Slice 2: Singleton Reader; root 0; parent 1; gang size 1; segment 1
 Slice 3: Singleton Reader; root 0; parent 1; gang size 1; segment 1
 Slice 4: Reader; root 0; parent 3; gang size 3
(28 rows)
```
**With patch:**
```
explain (slicetable,costs off) with cte as (insert into t_hashed select i, i*2 from generate_series(1,5) i  returning *),
cte2 as (select * from t_repl)
select * from cte join cte2 a on cte.i=a.i join (select t2.i from cte2 b join t2 using (i) limit 1) x on x.i=cte.i;
                                          QUERY PLAN                                           
-----------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)
   ->  Hash Join
         Hash Cond: (share0_ref2.i = t_hashed.i)
         ->  Shared Scan (share slice:id 1:0)
         ->  Hash
               ->  Hash Join
                     Hash Cond: (t_hashed.i = t2.i)
                     ->  Insert on t_hashed
                           ->  Redistribute Motion 1:3  (slice2; segments: 1)
                                 Hash Key: i.i
                                 ->  Function Scan on generate_series i
                     ->  Hash
                           ->  Broadcast Motion 1:3  (slice3; segments: 1)
                                 ->  Limit
                                       ->  Gather Motion 3:1  (slice4; segments: 3)
                                             ->  Limit
                                                   ->  Hash Join
                                                         Hash Cond: (share0_ref1.i = t2.i)
                                                         ->  Shared Scan (share slice:id 4:0)
                                                               ->  Seq Scan on t_repl t_repl_1
                                                         ->  Hash
                                                               ->  Seq Scan on t2
 Optimizer: Postgres-based planner
 Slice 0: Dispatcher; root 0; parent -1; gang size 0
 Slice 1: Reader; root 0; parent 0; gang size 3
 Slice 2: Singleton Reader; root 0; parent 1; gang size 1; segment 2
 Slice 3: Singleton Reader; root 0; parent 1; gang size 1; segment 2
 Slice 4: Reader; root 0; parent 3; gang size 3
(28 rows)

```

bimboterminator1 committed:
Fix wrong slice allocation for shared CTEs with modifying DML inside.

When the shared CTE with modifying DML inside is planned, at the first
occurrence of the CTE create_ctescan_plan function constructs the subplan, so
it's slice gets the gang type GANGTYPE_PRIMARY_WRITER. For the other occurrences
the subplan construction is being skipped, so if they are located in other
slices, and the CTE is modifying, their slices may not have
GANGTYPE_PRIMARY_WRITER gang type. Later, the ShareInputScans are assigned the
producer and consumer roles by apply_shareinput_dag_to_tree function, which
uses tree walker with specific traverse order. There was a case when the
consumer was created at GANGTYPE_PRIMARY_WRITER slice, while the producer were
put to the GANGTYPE_PRIMARY_READER slice. This lead to execution error when the
reader gang attempted to execute ModifyTable node.

This patch fixed the issue the following way. When creating the shared plan at
the first time, the gangType of root->curSlice is checked. If it has switched
compared to the value before the planning, we mark ShareInputScan as writer
slice maker (the rootSliceIsWriter flag is set to true). At other occurences of
CTE their ShareInputScan nodes are also marked with this state. After the
consumer, producer roles are assigned to each ShareInputScan, we set the
correct gang type to the producer (GANGTYPE_PRIMARY_WRITER), and to the
consumers (GANGTYPE_PRIMARY_READER, if the consumer is in different slice from
producer) in shareinput_mutator_xslice_2() function.

RekGRpth commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -993,12 +993,12 @@ shareinput_mutator_xslice_2(Node *node, PlannerInfo *root, bool fPop)
 		sisc->producer_slice_id = pershare->producer_slice_id;
 		sisc->nconsumers = bms_num_members(pershare->participant_slices) - 1;
 
+		PlanSlice  *currentSlice = &ctxt->slices[motId];
```
Mixing declarations and code!

bimboterminator1 replied:
Fixed

RekGRpth commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -1014,6 +1014,25 @@ shareinput_mutator_xslice_2(Node *node, PlannerInfo *root, bool fPop)
+		 * Therefore, in order to prevent this, we set up the correct
+		 * gangType back.
+		 */
+		if (plan->lefttree && sisc->rootSliceIsWriter)
```
```suggestion
		if (sisc->rootSliceIsWriter)
		{
			if (plan->lefttree)
				currentSlice->gangType = GANGTYPE_PRIMARY_WRITER;
			else if (sisc->this_slice_id != sisc->producer_slice_id)
				currentSlice->gangType = GANGTYPE_PRIMARY_READER;
		}
```

bimboterminator1 replied:
Fixed

RekGRpth commented on src/backend/optimizer/plan/createplan.c:
```diff
@@ -4394,13 +4394,30 @@ create_ctescan_plan(PlannerInfo *root, Path *best_path,
+			 */
+			if (root->curSlice->gangType != saved_gangType &&
+				root->curSlice->gangType == GANGTYPE_PRIMARY_WRITER)
+				cteplaninfo->rootSliceIsWriter = true;
```
```suggestion
				cteplaninfo->shared_plan->rootSliceIsWriter = true;
```

bimboterminator1 replied:
It doesnt work that way. `shared_plan` is Plan node, which is put under ShareInputScan in `share_prepared_plan`. And ShareInputScan is unique for each CTE refernce. In this conditions we need somehow to mark each ShareInputScan corresponding to the same CTE

RekGRpth replied:
1) `cteplaninfo->shared_plan == NULL`
You do: `cteplaninfo->rootSliceIsWriter = true;`
I suggest: `cteplaninfo->shared_plan->rootSliceIsWriter = true;`
2) next `cteplaninfo->shared_plan != NULL`
You check: `if (cteplaninfo->rootSliceIsWriter)`
I suggest: `if (cteplaninfo->shared_plan->rootSliceIsWriter)`
What is the difference?

bimboterminator1 replied:
`cteplaninfo->shared_plan` is a Plan node, which is shared between ShareInputScans, each ShareInputScan is unique, and is created  as ` subplan = share_prepared_plan`. I need to know somehow about the state of other ShareInputScan, but i have common access only to `cteplaninfo` or `cteplaninfo->shared_plan`.
Instead of `cteplaninfo` struct I will have to put `rootSliceIsWriter` to `Plan` struct (and support serialization for that field).

RekGRpth replied:
OK, I see

RekGRpth commented on src/backend/optimizer/plan/createplan.c:
```diff
@@ -4394,13 +4394,30 @@ create_ctescan_plan(PlannerInfo *root, Path *best_path,
 		/* Wrap the common Plan tree in a ShareInputScan node */
 		subplan = share_prepared_plan(cteroot, cteplaninfo->shared_plan);
+
+		if (cteplaninfo->rootSliceIsWriter)
```
```suggestion
		if (cteplaninfo->shared_plan->rootSliceIsWriter)
```

bimboterminator1 replied:
It doesnt work that way. shared_plan is Plan node, which is put under ShareInputScan in share_prepared_plan. And ShareInputScan is unique for each CTE refernce. In this conditions we need somehow to mark each ShareInputScan corresponding to the same CTE

RekGRpth commented on src/include/nodes/pathnodes.h:
```diff
@@ -490,6 +490,15 @@ typedef struct CtePlanInfo
+	 * the shared plan. This flag is needed for correct gangType allocation
+	 * after consumer's plan is cut.
+	 */
+	bool rootSliceIsWriter;
```
```suggestion
```

bimboterminator1 replied:
It doesnt work that way. shared_plan is Plan node, which is put under ShareInputScan in share_prepared_plan. And ShareInputScan is unique for each CTE refernce. In this conditions we need somehow to mark each ShareInputScan corresponding to the same CTE

bimboterminator1 committed:
Refactor gang type assignment

RekGRpth commented on src/test/regress/sql/with.sql:
```diff
@@ -1190,3 +1190,17 @@ create temp table with_test (i int);
 select * from with_test;
 drop table with_test;
+
+-- Test that planner correctly assigns writer and reader slices in case of
```
I run test and got such diff
```diff
diff -I HINT: -I CONTEXT: -I GP_IGNORE: -U3 /home/src/gpdb7/src/test/regress/expected/with_optimizer.out /home/src/gpdb7/src/test/regress/results/with.out
--- /home/src/gpdb7/src/test/regress/expected/with_optimizer.out    2023-12-27 14:47:50.070179849 +0500
+++ /home/src/gpdb7/src/test/regress/results/with.out   2023-12-27 14:47:50.114180633 +0500
@@ -2460,9 +2460,9 @@
                                              ->  Result
  Slice 0: Dispatcher; root 0; parent -1; gang size 0
  Slice 1: Reader; root 0; parent 0; gang size 3
- Slice 2: Reader; root 0; parent 1; gang size 3
- Slice 3: Primary Writer; root 0; parent 1; gang size 3
- Slice 4: Singleton Reader; root 0; parent 3; gang size 1; segment 1
+ Slice 2: Primary Writer; root 0; parent 1; gang size 3
+ Slice 3: Reader; root 0; parent 1; gang size 3
+ Slice 4: Singleton Reader; root 0; parent 3; gang size 1; segment 0
 GP_IGNORE:(21 rows)

 DROP TABLE withmodcte;
```

bimboterminator1 replied:
Well, I see. It will be more reliable to test only execution in that case.
gpdb7 is upstream version? In my build I ran tests from adb-7.0.0 branch.
Moreover, there aren't any CI tests for our 7.x  branch, and I doubt that we will start merging tickets before CI support.

RekGRpth replied:
> gpdb7 is upstream version?

No! I just run test on ADBDEV-4345 branch

RekGRpth replied:
Why does the same (Postgres-based) planner produce different results?

bimboterminator1 replied:
If I run the whole test suit from parallel_schedule (before the group with `with` test), all the tests pass correctly.

RekGRpth replied:
I run so:
```sh
./pg_regress --load-extension=gp_inject_fault --init-file=init_file create_function_1 create_type create_table int8 int4 copy with
```

RekGRpth replied:
hmm...
```diff
diff -I HINT: -I CONTEXT: -I GP_IGNORE: -U3 /home/src/gpdb7/src/test/regress/expected/with.out /home/src/gpdb7/src/test/regress/results/with.out
--- /home/src/gpdb7/src/test/regress/expected/with.out  2023-12-27 16:09:01.065718112 +0500
+++ /home/src/gpdb7/src/test/regress/results/with.out   2023-12-27 16:09:01.109718736 +0500
@@ -2460,9 +2464,9 @@
                                              ->  Result
  Slice 0: Dispatcher; root 0; parent -1; gang size 0
  Slice 1: Reader; root 0; parent 0; gang size 3
- Slice 2: Reader; root 0; parent 1; gang size 3
- Slice 3: Primary Writer; root 0; parent 1; gang size 3
- Slice 4: Singleton Reader; root 0; parent 3; gang size 1; segment 2
+ Slice 2: Primary Writer; root 0; parent 1; gang size 3
+ Slice 3: Reader; root 0; parent 1; gang size 3
+ Slice 4: Singleton Reader; root 0; parent 3; gang size 1; segment 1
 GP_IGNORE:(21 rows)

 DROP TABLE withmodcte;
```

bimboterminator1 replied:
At least `segment 2/1`  part is flacky. It means that we need either add `matchsubs` , or just test the execution only

RekGRpth replied:
what about order of `Slice 2: Primary Writer` and `Slice 3: Reader` ?

bimboterminator1 replied:
I can't say anything without a plan.

RekGRpth replied:
plans are same (as in expected files)

bimboterminator1 replied:
The algorithm is idepended from enviroment, I'm getting correct plan, and the plan creation procedure is not ambiguous. 
I have the following diff on the same command (flacky part) or don't have diff at all (everything on ADBDEV-4345 branch):
```diff
--- /home/bimbo/Downloads/test_repo/src/test/regress/expected/with_optimizer.out	2023-12-27 14:25:56.010010388 +0300
+++ /home/bimbo/Downloads/test_repo/src/test/regress/results/with.out	2023-12-27 14:25:56.040010458 +0300
@@ -2462,7 +2462,7 @@
  Slice 1: Reader; root 0; parent 0; gang size 3
  Slice 2: Reader; root 0; parent 1; gang size 3
  Slice 3: Primary Writer; root 0; parent 1; gang size 3
- Slice 4: Singleton Reader; root 0; parent 3; gang size 1; segment 1
+ Slice 4: Singleton Reader; root 0; parent 3; gang size 1; segment 2
 GP_IGNORE:(21 rows)
```

bandetto committed:
Merge branch 'adb-7.1.0' into ADBDEV-4345

bandetto committed:
Update tests

bandetto committed:
Merge branch 'adb-7.1.0' into ADBDEV-4345

bandetto committed:
Fix tests

bandetto committed:
Fix segment number in tests

RekGRpth commented on src/test/regress/sql/with.sql:
```diff
@@ -1203,3 +1203,79 @@ WITH cte AS (
+SELECT i FROM cte a
+JOIN cte b USING(i);
+
+-- Case 2
```
Why do tests 2-5 pass without a patch?

bandetto committed:
Simplify tests back

bandetto committed:
Fix tests

bandetto committed:
Space

RekGRpth commented:
Why does analyze show never executed?
```sql
CREATE TABLE with_test (i int, j int) DISTRIBUTED RANDOMLY;
SET gp_cte_sharing = 1;

EXPLAIN (ANALYZE, SLICETABLE)
WITH cte AS (
	INSERT INTO with_test VALUES (1, 2) RETURNING *
)
SELECT i FROM cte a JOIN cte b USING (i);
                                                                 QUERY PLAN                                                                  
---------------------------------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.16..0.34 rows=3 width=4) (actual time=2.097..2.301 rows=1 loops=1)
   ->  Hash Join  (cost=0.16..0.30 rows=1 width=4) (never executed)
         Hash Cond: (a.i = b.i)
         ->  Redistribute Motion 3:3  (slice2; segments: 3)  (cost=0.00..0.12 rows=3 width=4) (never executed)
               Hash Key: a.i
               ->  Subquery Scan on a  (cost=0.00..0.06 rows=3 width=4) (never executed)
                     ->  Shared Scan (share slice:id 2:0)  (cost=0.03..0.23 rows=1 width=8) (never executed)
         ->  Hash  (cost=0.12..0.12 rows=3 width=4) (never executed)
               ->  Redistribute Motion 3:3  (slice3; segments: 3)  (cost=0.00..0.12 rows=3 width=4) (never executed)
                     Hash Key: b.i
                     ->  Subquery Scan on b  (cost=0.00..0.06 rows=3 width=4) (never executed)
                           ->  Shared Scan (share slice:id 3:0)  (cost=0.03..0.23 rows=1 width=8) (never executed)
                                 ->  Insert on with_test  (cost=0.00..0.03 rows=1 width=8) (never executed)
                                       ->  Redistribute Motion 1:3  (slice4; segments: 1)  (cost=0.00..0.03 rows=1 width=8) (never executed)
                                             ->  Result  (cost=0.00..0.01 rows=1 width=8) (never executed)
 Optimizer: Postgres-based planner
 Planning Time: 12.163 ms
 Slice 0: Dispatcher; root 0; parent -1; gang size 0
 Slice 1: Reader; root 0; parent 0; gang size 3
 Slice 2: Reader; root 0; parent 1; gang size 3
 Slice 3: Primary Writer; root 0; parent 1; gang size 3
 Slice 4: Singleton Reader; root 0; parent 3; gang size 1; segment 0
   (slice0)    Executor memory: 51K bytes.
   (slice1)    
   (slice2)    
   (slice3)    
   (slice4)    
 Memory used:  128000kB
 Execution Time: 11.123 ms
(29 rows)

WITH cte AS (
	INSERT INTO with_test VALUES (1, 2) RETURNING *
)
SELECT i FROM cte a JOIN cte b USING (i);
 i 
---
 1
(1 row)
```

bimboterminator1 commented:
> Why does analyze show never executed?

Probably because of the strategy:
https://github.com/arenadata/gpdb/pull/557

bandetto commented:
> Why does analyze show never executed?

```sql
EXPLAIN (ANALYZE, SLICETABLE)
WITH cte AS (
        INSERT INTO with_test VALUES (1) RETURNING *
)
SELECT i FROM cte;
                                                      QUERY PLAN                                                      
----------------------------------------------------------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..0.07 rows=3 width=4) (actual time=9.510..9.514 rows=1 loops=1)
   ->  Insert on with_test  (cost=0.00..0.03 rows=1 width=4) (never executed)
         ->  Result  (cost=0.00..0.01 rows=1 width=4) (never executed)
 Optimizer: Postgres-based planner
 Planning Time: 5.980 ms
 Slice 0: Dispatcher; root 0; parent -1; gang size 0
 Slice 1: Primary Writer; root 0; parent 0; gang size 1
   (slice0)    Executor memory: 36K bytes.
   (slice1)    
 Memory used:  128000kB
 Execution Time: 10.138 ms
(11 rows)
```

`(never executed)` shows up even without the patch on working DML queries. With diff:
```diff
diff --git a/src/backend/tcop/pquery.c b/src/backend/tcop/pquery.c
index 251860682e..48b17d4ded 100644
--- a/src/backend/tcop/pquery.c
+++ b/src/backend/tcop/pquery.c
@@ -375,10 +375,12 @@ ChoosePortalStrategy(List *stmts)
 					pstmt->copyIntoClause == NULL &&
 					pstmt->refreshClause == NULL)
 				{
-					if (pstmt->hasModifyingCTE)
-						return PORTAL_ONE_MOD_WITH;
-					else
+					if (!pstmt->hasModifyingCTE)
 						return PORTAL_ONE_SELECT;
+					else if (Gp_role != GP_ROLE_EXECUTE)
+						return PORTAL_ONE_MOD_WITH;
+
+					return PORTAL_MULTI_QUERY;
 				}
 				if (pstmt->commandType == CMD_UTILITY)
 				{
```

Stats are back:
```sql
EXPLAIN (ANALYZE, SLICETABLE)
WITH cte AS (
        INSERT INTO with_test VALUES (1) RETURNING *
)
SELECT i FROM cte;
                                                      QUERY PLAN                                                      
----------------------------------------------------------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..0.07 rows=3 width=4) (actual time=2.382..2.386 rows=1 loops=1)
   ->  Insert on with_test  (cost=0.00..0.03 rows=1 width=4) (actual time=0.213..0.214 rows=1 loops=1)
         ->  Result  (cost=0.00..0.01 rows=1 width=4) (actual time=0.012..0.012 rows=1 loops=1)
 Optimizer: Postgres-based planner
 Planning Time: 26.801 ms
 Slice 0: Dispatcher; root 0; parent -1; gang size 0
 Slice 1: Primary Writer; root 0; parent 0; gang size 1
   (slice0)    Executor memory: 36K bytes.
   (slice1)    Executor memory: 38K bytes (seg1).
 Memory used:  128000kB
 Execution Time: 3.348 ms
(11 rows)

```


Please look at this patch:
> Probably because of the portal strategy:
> https://github.com/arenadata/gpdb/pull/557

RekGRpth commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -1014,6 +1014,23 @@ shareinput_mutator_xslice_2(Node *node, PlannerInfo *root, bool fPop)
 					break;
 			}
 		}
+
+		/*
+		 * The SISC's plan contains modifying operation, which
+		 * creates a writer gang. Due to specific tree traverse order
+		 * during apply_shareinput_dag_to_tree, the producer could get
+		 * to the reader slice, while the consumer could get to the
+		 * writer slice (gangType is chosen before shareinput fixing).
+		 * Therefore, in order to prevent this, we set up the correct
+		 * gangType back.
+		 */
+		if (sisc->rootSliceIsWriter)
+		{
+			if (plan->lefttree)
+				currentSlice->gangType = GANGTYPE_PRIMARY_WRITER;
+			else if (sisc->this_slice_id != sisc->producer_slice_id)
+				currentSlice->gangType = GANGTYPE_PRIMARY_READER;
+		}
```
I think we should move this before above switch

bandetto committed:
Move gang reassignment before the switch

Commit:
Merge branch 'adb-7.1.0' into ADBDEV-4345
